### ISSUE TITLE
Clarify that construct effective {content,processor} profile is performed only once.

### ISSUE THREAD

**Issue by skynavga** [w3c/ttml2#860]
Add notes in §5.2.4.2 and §5.2.4.3, under the definitions of **[construct effective content profile]** and **[construct effective processor profile]**, respectively, that these procedures are performed only once during the life cycle of a document, i.e., that if they are invoked more than once, then subsequent invocations obtain the results produced by the first invocation (the values are memoized).

**Comment by nigelmegitt** [comment]
> only once during the life cycle of a document

That can not be true (because it is too general) - they're _probably_ performed only once during a particular processor _instance_'s processing of a single document instance, but that might be an implementation feature, for example, a processor that simulates the effect of processing a document against a set of profile constraints could reprocess the same document instance multiple times, calling those procedures each time.

Suggest closing and taking no action here.

**Comment by skynavga** [comment]
But that is the definition of the life cycle of a document: a particular processor instance's processing of a single document instance.

**Comment by nigelmegitt** [comment]
That doesn't seem to be the common usage definition of the life cycle of a document, which would begin at conception, then authoring/transcoding, distribution, finally processing. Do we have a narrower definition written down somewhere?

**Comment by skynavga** [comment]
s/document life cycle/document instance processing life cycle/

**Comment by nigelmegitt** [comment]
Even then I think it is an unnecessary note - it is either obvious (for 99% of cases) or wrong (for the exceptions that will surely exist).

**Comment by skynavga** [comment]
The note is important to communicate expectations to readers, whether
authors or implementors. I will formulate the note in such a way as to
handle implementation specific exceptions, but there are no exceptions
driven by normative specification language.

**Comment by skynavga** [comment]
Another way to state this same comment is that the values of ECP and EPP, and the objects they reference, never change once they are set to a non-null value. This is actually an unstated assumption in the specification, and important to document, at a minimum, on a non-normative basis.

**Comment by skynavga** [comment]
Regarding https://github.com/w3c/ttml2/issues/860#issuecomment-400254374, it is apparently not obvious since you have imagined a scenario where it might not hold (which I would argue is already outside the scope of the specification), and it is certainly not wrong, since many semantics of the specification assume it to be true (without stating as such).

**Comment by nigelmegitt** [comment]
> it is certainly not wrong, since many semantics of the specification assume it to be true 

Can you point to specific examples?

**Comment by nigelmegitt** [comment]
> Note that document editing applications invalidate all assumptions (during the editing process), so are out of scope in general.

I don't think you can sweep authoring tools out of the equation like this; similarly validation tools would also be in the class of processor that might process a document instance either multiple times sequentially or multiple times in parallel to establish their validity against multiple ECPs or EPPs. You might argue that in those cases each processing pass is a different instance of the document instance processing lifecycle, which is fine, but just proves my point.

That point being that the note is either obvious or misleading or unnecessarily constraining. I'm happy to add this to the agenda but right now I'm unconvinced that the spec would be improved by adding the note; in fact it looks like it would be detrimental. Implementers can make their own decisions about this. 

You might want to construct a "document instance processing session" abstraction that has a one-time-evaluated ECP and EPP and a document processing context, in which case I'd agree that the note would apply to that session, but constraining all implementations to operate a single document instance processing session is going too far.

**Comment by skynavga** [comment]
@nigelmegitt we have a fundamental disagreement then

**Comment by nigelmegitt** [comment]
@skynavga Are you saying it is impossible for an implementation to process a single document instance against >1 EPPs in the same number of passes as it takes to process it for 1 EPP?

**Comment by skynavga** [comment]
@nigelmegitt i'm saying that generic processor conformance assumes that EPP doesn't change during processing, the reason being is that [validate document] and [abort if unsupported processor profile] are only executed once; if EPP could change, then both of these would have to be performed again

**Comment by nigelmegitt** [comment]
@skynavga I think you've argued my case - in the case that an implementation wants to execute those sub-processes only once they must not change EPP during processing; implementations that can execute them multiple times are not held by the same constraint. The spec doesn't need to make the assumption one way or the other.

**Comment by skynavga** [comment]
@nigelmegitt not at all, an implementation that would allow EPP to vary is out of scope; the current spec as written assumes it does not change; saying that this is the case does not prevent creating implementations that go outside the defined scope

**Comment by nigelmegitt** [comment]
I've reviewed the recent commits and they still don't add up to something that I feel is an improvement.

> an implementation that would allow EPP to vary is out of scope

@skynavga Aside from the point that I don't recall us ever agreeing that, this doesn't actually address the point I made which is that a (static) document instance can be processed against multiple EPPs during parallel or sequential processing. The issue is seeking to freeze into the spec assumptions about processing that are too strict.

My position here is that I don't accept the issue as formulated and do not wish to make the change. As such, right now, we don't have consensus.

I did hint or propose a way to factor this into something that I could agree to, at https://github.com/w3c/ttml2/issues/860#issuecomment-402153540, but so far you have neither disagreed with that nor adopted it.

**Comment by skynavga** [comment]
@nigelmegitt regarding

>a (static) document instance can be processed against multiple EPPs during parallel or sequential processing

I'm not sure what is meant by "parallel or sequential processing". If you are suggesting that a single document instance be subdivided into segments which are processed in parallel or serially (or both), then the processing of each individual segment would be the scope of constant EPP. In particular, the process of segmentation is outside the scope of TTML and only the segments, which would individually be treated as distinct TTML document instances, would be in scope for TTML semantics, including EPP constancy.

**Comment by nigelmegitt** [comment]
> what is meant by "parallel or sequential processing"

I'm envisaging a validating processor doing a single pass through a single document instance evaluating it simultaneously against a set of more than one "hypothetical" EPPs to test their validity against each EPP.

**Comment by skynavga** [comment]
@nigelmegitt ok, but that would be doing something different than is currently defined, making it out of scope; i.e., the current definition determines the EVP (effective validation profile) exactly one time before validation, and uses that determination alone; we have defined no process that speculatively attempts validating against multiple EVPs

**Comment by nigelmegitt** [comment]
@skynavga We have neither admitted nor rejected it, and by introducing the proposed new text we begin to hint at a processing model that incorporates assumptions that we have not validated. We do not need to do that.

**Comment by skynavga** [comment]
@nigelmegitt just because we have not explicitly documented something does not mean we do (did) not intend something to be such and such; it is our position that constancy of ECP/EPP/EVP indeed **holds** for the purpose of performing the sequence of processing defined by TTML on a single document instance; I don't mind documenting your speculative implementation as an out of scope exception to such constancy, but the default is constancy

**Comment by skynavga** [comment]
Merged early per WG resolution to move forward with CR3 CfC.
